# Background
This repository contains the code to calculate genomic coverage from an NGS BAM file generated from whole genome sequencing (WGS) data aligned against the Genome in a Bottle (GIAB, NA12878/HG001) reference sample. The sample was aligned to GRCh38 reference genome and the coverage was calculated across chromosomes instead of per base position / gene / CpG etc... as a basic example.

# Walkthrough

## BAM file download and set-up
- A short BAM file (1M reads) was used an example in this repo
    - Full BAM (16G / ~170M reads) downloaded with `wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/NA12878/alignment/NA12878.mapped.ILLUMINA.bwa.CEU.low_coverage.20121211.bam`


## BED file download and set-up
- Initial BED file was generated by `cat ./GCF_000001405.26_GRCh38_genomic.gff | gff2bed > ./GCF_000001405.26_GRCh38_genomic.bed`
    - This used the bedops suite of tools
    - Comments may need to be removed from file first

- The chromosome reference names were then replaced with the chromosome number using the command `./parse_chroms.sh -b ./GCF_000001405.26_GRCh38_genomic.bed`

## Calculating coverage
- To download the code required for the `./src/python/calc_target_coverage.py` script, run `pip install -r requirements.txt`
- Once the dependencies are set up, the script can be run using `./src/python/calc_target_coverage.py -a ./NA12878.mapped.ILLUMINA.bwa.CEU.low_coverage.20121211.sub.bam -b ./chromosomes.bed --threads 6`
    - The help message for the script is shown below:
```
usage: calc_target_coverage.py [-h] [-a ALIGNMENT] [-b BED] [-t THREADS]

Script to calculate coverage across regions from a target BED file

options:
  -h, --help            show this help message and exit
  -a ALIGNMENT, --alignment-file ALIGNMENT
                        Path to alignment file (SAM / BAM)
  -b BED, --bed-file BED
                        Path to bed file
  -t THREADS, --threads THREADS
                        Number of threads to use for parallel processing
```

- The results will be printed to STDOUT and will look like the below:
```
Chromosome 1: 999,897
Chromosome 2: 0
Chromosome 3: 0
Chromosome 4: 0
Chromosome 5: 0
Chromosome 6: 0
Chromosome 7: 0
Chromosome 8: 0
Chromosome 9: 0
Chromosome 10: 0
Chromosome 11: 0
Chromosome 12: 0
Chromosome 13: 0
Chromosome 14: 0
Chromosome 15: 0
Chromosome 16: 0
Chromosome 17: 0
Chromosome 18: 0
Chromosome 19: 0
Chromosome 20: 0
Chromosome 21: 0
Chromosome 22: 0
Chromosome X: 0
Chromosome Y: 0
```

- NOTE: The 'sub' BAM file takes the first 1M reads and they appear to be sorted hence all of the reads coming just from the first chromosome
    - Using the full BAM (too large to add to the repo) yields the below results:
```
Chromosome 1: 12,892,698
Chromosome 2: 13,747,916
Chromosome 3: 11,334,125
Chromosome 4: 11,004,704
Chromosome 5: 10,405,139
Chromosome 6: 9,780,726
Chromosome 7: 8,870,298
Chromosome 8: 8,333,172
Chromosome 9: 6,767,853
Chromosome 10: 7,287,394
Chromosome 11: 7,429,084
Chromosome 12: 7,430,177
Chromosome 13: 5,564,465
Chromosome 14: 5,059,702
Chromosome 15: 4,559,245
Chromosome 16: 4,849,844
Chromosome 17: 4,170,905
Chromosome 18: 4,316,255
Chromosome 19: 2,802,277
Chromosome 20: 3,245,545
Chromosome 21: 2,020,211
Chromosome 22: 1,761,554
Chromosome X: 8,837,533
Chromosome Y: 42,759
```
# Testing
- Some basic tests were added to `./testing/python/` to test the different aspects of code for calculating the coverages
- Test code can be ran by FIRST going navigating into the `testing/python/` directory and then simply typing `pytest`
